Fix instantiating unikernels and tailing logs

nderjung opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This makes several changes across subsystems to fix the instantiation and tailing of unikernels via `kraft run`. A number of changes were required, particularly:

 - Deprecating the use of a UNIX socket for the logs of a QEMU instance -- since there could be a synchronization issue if the consumer was not fast enough to start receiving logs. Instead, the logs are written verbatim to log file which itself has been made available in `machine.MachineConfig`. Necessary adjustments are made to facilitate "tailing" from a log file from QEMU. This has the added benefit being able to retrieved after the unikernel has ended its life.
 - Fixing the instantiation, tailing and monitoring of a unikernel in `kraft run`. The instantiation of the events monitor is made to be completely silent such that sub-process noise does not spill over into the executor's parent's stdout. The actual instantiation of the unikernel via `driver.Start()` is moved as late-as-possible such that go-routines can be established first. Additionally, the manipulation of the cancalable-context is fixed with respect to the fixed `TailWriter`.
 - Introduce an optional CLI argument for setting the name of the unikernel;
 - Introduce an optional CLI argument for setting the initrd (CPIO) file for the unikernel.

### How to test

I used a simple "Hello, World" C unikernel, with 1 target (kvm-x86_64) and a simple loop in `main.c`, e.g.:

```c
#include <stdio.h>
#include <time.h>
#include <errno.h>

static void millisleep(unsigned int millisec)
{
        struct timespec ts;
        int ret;

        ts.tv_sec = millisec / 1000;
        ts.tv_nsec = (millisec % 1000) * 1000000;
        do
                ret = nanosleep(&ts, &ts);
        while (ret && errno == EINTR);
}


int main(int argc, char *argv[])
{
        for (;;) {
                printf("Hello world!\n");
                millisleep(1000);
        }

        return 0;
}
```

Once built as a unikernel, I could do the following:

```
# this will instantiate `hello1` and immediately detach.
kraft run --detach --name hello1

# i can now see there `kraft events` and `qemu-system-x86_64` running in the background

# starting another unikernel will start to tail the logs
kraft run --name hello2

# Ctrl+C will stop tailing

# i can kill them both
kraft stop --all

# see that they are both exited
kraft ps

# clean up
kraft rm --all

# no lingering process should be on the system (aka `kraft events [...]`)
```

nderjung committed:
feat(machine): Add `LogFile` as attribute to `MachineConfig`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(qemu): Set the default path of the machine's `LogFile`

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(qemu): tail logs from file

This commit fixes the "tailing" of machine logs from QEMU by forcing
the serial console into a log file and then using fsnotify to receive
updates of file writes to continiously write the latest file contents.

This has the added benefit of porcelaining the logs into the runtime
dir (/var/kraftkit) for later retrieval.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(qemu): fix spelling mistake

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
Squash the license and copyright notice preamble

This commit "squashes" the license and copyright preamble in files
associated with this sign off.  The preamble already includes the SPDX
license identifier which essentially duplicates the text in question.
Additionally, this license, in accordance with the license, is
distributed with the source code via the `LICENSE.md` file which is
accompanied in this repository.

Additionally, copyright notice is provided to The KraftKit Authors,
which entitles new developers who contribute commits to the project an
associated copyright with regard to their contributions without having
to specifically update the license and copyright preamble.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(events): use internal waitgroup system

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(waitgroup): introduce a generic wait group system

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(run): remove unused `Volumes` command-line flag

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(events): correct spelling mistakes

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(events): correctly handle exit events with `--quit-together`

This commit fixes the functionality of the events program which was not
exiting after all unikernel virtual machines had exited.  Instead the
process lingered.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(events): convert warning to debug log message

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(events): remove spurious colon

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): set the correct usage syntax

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): set the correct number of arguments

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): correctly handle the execution order of input entities

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(run): add optional `-n|--name` command-line option

This feature adds the ability to specifically set the name of the
machine instance from the command-line.  If not provided, the default
randomly generated name will be used.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(run): add optional `-i|--initrd` command-line flag

This feature allows setting a path to an optional CPIO file for the
unikernel.

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(run): switch log message to debug-level

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(run): correctly handle executing and tailing logs of a unikernel

This commit fixes the instantiation of a unikernel, tailing its logs and
preventing the execution of the spawned process (`kraft events`) which
did not correctly handle the facilitation of unikernel exits.

First, the instantiation of the events monitor is removed from `kraft
run`.  The original purpose of this hidden subcommand was to aid in
removing or destroying of a unikernel.  Instead this functionality is
added directly to `kraft run`.

Secondly, the actual instantiation of the unikernel via `driver.Start()`
is moved as late-as-possible such that go-routines can be established
first.  The removal or destruction of a unikernel is affixed at the end
of the execution such that we can wait for an exit event.  Having the
event within the context of a Go routine would prematurely exit the
program and thus the unikernel could not be stopped or removed.

Finally, manipulation of the cancelable-context is fixed with
respect to the fixed `TailWriter`.

Signed-off-by: Alexander Jung <alex@unikraft.io>

antoineco reviewed (approved):
Looking great!

craciunoiuc reviewed (changes requested):
Some things from my side:

```
 1. `kraft stop X` - does not also clean the `.pid` - this makes subsequent calls to `kraft ps` fail.
 2. Ctrl^C should not work and Ctrl^D should stop the unikernel.
 3. `kraft stop/rm --all` should exist
 4. `kraft rm X` only removes an unikernel from the tracked list - if you forget to run `kraft stop` before it is lost forever - `kraft rm` should ask the user if they want to also stop
 5. `kraft events` should have a configurable polling rate and by default poll once a second/half-second
 6. `kraft ps` should run in less than one second
 7. killing an unikernel should make it eventually be removed from `kraft ps` -> 30s timeout maybe?
 8. `kraft stop/rm/etc.` commands should auto-complete ids when they can (so you don't have to input the entire id)
```

In commit: `fix(qemu): tail logs from file`: `porclaining` ->`porcelaining`
In commit: `fix(run): correctly handle executing and tailing logs of a unikernel` `cancalable` -> `cancelable`

nderjung commented:
Thanks for the feedback @craciunoiuc, i'll take a look :)

nderjung committed:
fix(cmd/kraft): Spelling mistake in license header preamble

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
fix(stop): Do not check state before invoking remove

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(stop): Introduce `--all` flag

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(stop): Allow stopping by name

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(rm): Add `--all` flag

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(rm): Allow removing by name

Signed-off-by: Alexander Jung <alex@unikraft.io>

nderjung committed:
feat(logs): Introduce `kraft logs` subcommand

This allows the user to read and follow unikernel logs and has a similar
CLI interface to popular applications such as docker and kubectl.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (changes requested):
A few things from my side:

1. Running the `kraft run` command on helloworld, without `--rm` makes the unikernel stop before the kraft can check for it. This puts everything in an 'undefined' state where it just spams:
```
could not stop 1a079215293cce4bb0e7b4e088c847735ccef7563d20d4c88ace18e7b14a0257: dial unix /var/kraftkit/1a079215293cce4bb0e7b4e088c847735ccef7563d20d4c88ace18e7b14a0257_control.sock: connect: no such file or directory
```

Afterwards `kraft ps` and `kraft run` no longer work until you delete `/var/kraftkit`.

My suggestion is to treat the special case of fast exiting VMs, or just switch from `--rm` to `--no-rm`.

2. Should you be able to see logs from stopped machines?

3. Should machines with an exited state show up in in `kraft ps --all`? I can't seem to find them, probably because they exit too fast. Update: Even the ones that I can see no longer show afterwards:
```
➜  doom git:(staging) ✗ sudo ~/kraftkit/dist/kraft ps --all
MACHINE ID    IMAGE                     ARGS  CREATED         STATUS   MEM
838d86fcab74  kernel://doom_kvm-x86_64        10 seconds ago  running  0MB
➜  doom git:(staging) ✗ sudo ~/kraftkit/dist/kraft ps --all
MACHINE ID  IMAGE  ARGS  CREATED  STATUS  MEM
```
4. With `kraft run` the serial interface becomes uni-directional. I can no longer do `scanf` inside the code. This happens probably because it's writing to a file and then just spewing out the file to the console, so you can no longer interact.
Because of this, the unikernel appears to be hanging, but it's just waiting for input.

I think this is enough for now

nderjung committed:
fix(qemu): Correctly handle looking up "dead" instances

Before this change, looking up the state of a dead instance (one whose
QEMU userspace monitor has exited) would directly result in an error
before being able to return the (updated) state.

Signed-off-by: Alexander Jung <alex@unikraft.io>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
